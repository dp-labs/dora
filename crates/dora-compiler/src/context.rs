use melior::{
    Context as MLIRContext,
    dialect::DialectRegistry,
    utility::{register_all_dialects, register_all_llvm_translations, register_all_passes},
};
use std::path::PathBuf;

/// The `Context` struct represents the MLIR (Multi-Level Intermediate Representation)
/// context in which various operations, types, and other IR constructs exist.
/// It serves as the core environment for creating, managing, and manipulating
/// MLIR-based intermediate representations.
///
/// # Fields:
/// - `mlir_context`: The `MLIRContext` is the underlying context used to manage
///   all MLIR-related constructs, ensuring proper memory management and scoping
///   of IR elements.
///
/// # Example Usage:
/// ```no_check
/// let ctx = Context::new();
/// ```
///
/// This struct is essential when working with MLIR in a Rust-based environment,
/// providing the foundational context to operate on IR.
#[derive(Debug, Eq, PartialEq)]
pub struct Context {
    pub mlir_context: MLIRContext,
}

unsafe impl Send for Context {}
unsafe impl Sync for Context {}

impl Default for Context {
    fn default() -> Self {
        Self::new()
    }
}

impl Context {
    pub fn new() -> Self {
        let mlir_context = initialize_mlir();
        Self { mlir_context }
    }
}

/// The `Session` struct represents a session in which an MLIR file is processed.
/// It contains paths to the raw and after-pass MLIR files, allowing the user to track
/// different stages of the MLIR transformation process.
///
/// # Fields:
/// - `raw_mlir_path`: This field holds an optional `PathBuf` pointing to the path
///   of the raw MLIR file before any transformations or passes have been applied.
///   It is `None` if the raw MLIR file path is not provided.
/// - `after_pass_mlir_path`: This field holds an optional `PathBuf` pointing to the path
///   of the MLIR file after various passes and transformations have been applied.
///   It is `None` if the after-pass MLIR file path is not provided.
///
/// # Example Usage:
/// ```no_check
/// let session = Session {
///     raw_mlir_path: Some(PathBuf::from("input.mlir")),
///     after_pass_mlir_path: Some(PathBuf::from("output.mlir")),
/// };
/// ```
///
/// The `Session` struct helps to keep track of MLIR files at different stages
/// during compilation, especially in workflows where passes are applied to modify
/// the IR in successive steps.
#[derive(Debug, Clone, Default)]
pub struct Session {
    /// The path for the raw mlir file before any transformation passes.
    pub raw_mlir_path: Option<PathBuf>,
    /// The path for the MLIR file after transformation passes have been applied.
    pub after_pass_mlir_path: Option<PathBuf>,
}

/// Initializes and configures a new `MLIRContext` with all available dialects and passes.
///
/// This function sets up an `MLIRContext` by performing the following steps:
/// - Creates a new `MLIRContext`.
/// - Appends a registry of all available dialects.
/// - Attaches a diagnostic handler that logs diagnostics to `stderr`.
/// - Loads all available dialects into the context.
/// - Allows the use of unregistered dialects within the context.
/// - Registers all available optimization and transformation passes.
/// - Registers LLVM translations for further compilation steps.
///
/// # Returns
///
/// A fully initialized `MLIRContext` ready for compiling and optimizing MLIR-based operations.
///
/// # Behavior
///
/// - This context is configured to handle diagnostics and allows unregistered dialects to be used.
/// - It includes all MLIR dialects and passes available in the current setup.
/// - The context supports LLVM translations for lowering MLIR to LLVM IR.
pub fn initialize_mlir() -> MLIRContext {
    let context = MLIRContext::new();
    context.append_dialect_registry(&{
        let registry = DialectRegistry::new();
        register_all_dialects(&registry);
        registry
    });
    context.attach_diagnostic_handler(|diagnostic| {
        eprintln!("{diagnostic}");
        true
    });
    context.load_all_available_dialects();
    context.set_allow_unregistered_dialects(true);
    register_all_passes();
    register_all_llvm_translations(&context);
    context
}
