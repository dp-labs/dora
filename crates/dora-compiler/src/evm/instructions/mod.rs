use melior::ir::BlockRef;

use crate::conversion::builder::OpBuilder;
use crate::evm::EVMCompiler;
use crate::evm::backend::EVMBuilder;

use super::CtxType;

pub mod arithmetic;
pub mod bitwise;
pub mod contract;
pub mod control;
pub mod host;
pub mod host_env;
pub mod memory;
pub mod stack;
pub mod system;

impl<'c> EVMCompiler<'c> {
    #[inline]
    pub(crate) fn make_builder<'a>(
        ctx: &'a mut CtxType<'c>,
        block: BlockRef<'c, 'a>,
    ) -> EVMBuilder<'a, 'c> {
        EVMBuilder {
            builder: OpBuilder::new_with_block(ctx.context, block),
            ctx,
            use_static_stack: false,
        }
    }
}
