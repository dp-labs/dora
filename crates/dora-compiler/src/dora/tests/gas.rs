use crate::context::Context;
use crate::dora::gas::GasOptions;
use crate::evm::program::Operation;
use crate::evm::program::Program;
use crate::evm::CompileOptions;
use crate::evm::EVMCompiler;
use crate::Compiler;
use dora_primitives::spec::SpecId;
use num_bigint::BigUint;

macro_rules! assert_snapshot {
    ($operations:expr) => {
        let program = Program {
            operations: $operations,
            code_size: 0,
        };
        let context = Context::new();
        let compiler = EVMCompiler::new(&context);
        let mut module = compiler
            .compile(&program, &(), &CompileOptions::default())
            .expect("failed to compile program");

        crate::evm::pass::run(&context.mlir_context, &mut module.mlir_module).unwrap();
        crate::dora::pass::run_gas_pass(
            &context.mlir_context,
            &mut module.mlir_module,
            GasOptions {
                spec_id: SpecId::CANCUN,
                limit_contract_code_size: None,
            },
        )
        .unwrap();
        let op = module.module().as_operation();
        assert!(op.verify());
        insta::assert_snapshot!(op);
    };
}

#[test]
fn gas_push_push_add() {
    let (a, b) = (BigUint::from(3_u8), 3_u32);
    let operations = vec![
        Operation::Push((1_u8, BigUint::from(b))),
        Operation::Push((1_u8, a.clone())),
        Operation::Add,
    ];
    assert_snapshot!(operations);
}

#[test]
fn gas_push_push_exp() {
    let (a, b) = (2_u32, 3_u32);
    let operations = vec![
        Operation::Push((1_u8, BigUint::from(b))),
        Operation::Push((1_u8, BigUint::from(a))),
        Operation::Exp,
    ];
    assert_snapshot!(operations);
}

#[test]
fn gas_push_push_add_sub() {
    let (a, b) = (BigUint::from(3_u8), 3_u32);
    let operations = vec![
        Operation::Push((1_u8, BigUint::from(b))),
        Operation::Push((1_u8, a.clone())),
        Operation::Add,
        Operation::Push((1_u8, a.clone())),
        Operation::Sub,
        Operation::Push((2_u8, a.clone())),
        Operation::Mul,
    ];
    assert_snapshot!(operations);
}

#[test]
fn gas_call_data_copy() {
    let operations = vec![
        Operation::Push((1_u8, BigUint::from(32_u8))),
        Operation::Push((1_u8, BigUint::from(0_u8))),
        Operation::Push((1_u8, BigUint::from(0_u8))),
        Operation::CallDataCopy,
    ];
    assert_snapshot!(operations);
}

#[test]
fn gas_create() {
    let operations = vec![
        Operation::Push((1_u8, BigUint::from(41_u8))),
        Operation::Push((1_u8, BigUint::from(0_u8))),
        Operation::Push((1_u8, BigUint::from(0_u8))),
        Operation::Create,
    ];
    assert_snapshot!(operations);
}

#[test]
fn gas_call() {
    let operations = vec![
        Operation::Push((4_u8, BigUint::from(10000_u32))),
        Operation::Push((4_u8, BigUint::from(0x1000_u32))),
        Operation::Push((1_u8, BigUint::from(1_u32))),
        Operation::Push((1_u8, BigUint::from(32_u32))),
        Operation::Push((1_u8, BigUint::from(32_u32))),
        Operation::Push((1_u8, BigUint::from(64_u32))),
        Operation::Push((1_u8, BigUint::from(64_u32))),
        Operation::Call,
    ];
    assert_snapshot!(operations);
}
