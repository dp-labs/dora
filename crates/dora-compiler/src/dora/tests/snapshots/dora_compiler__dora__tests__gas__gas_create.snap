---
source: crates/dora-compiler/src/dora/tests/gas.rs
expression: op
snapshot_kind: text
---
module {
  llvm.mlir.global internal @dora_global_stack_ptr() {addr_space = 0 : i32} : !llvm.ptr
  llvm.mlir.global internal @dora_global_stack_length() {addr_space = 0 : i32} : i64
  llvm.mlir.global internal @dora_global_memory_ptr() {addr_space = 0 : i32} : !llvm.ptr
  llvm.mlir.global internal @dora_global_memory_size() {addr_space = 0 : i32} : i64
  llvm.mlir.global internal @dora_global_calldata_ptr() {addr_space = 0 : i32} : !llvm.ptr
  llvm.mlir.global internal @dora_global_calldata_size() {addr_space = 0 : i32} : i64
  llvm.mlir.global internal @dora_global_gas_counter() {addr_space = 0 : i32} : i64
  llvm.mlir.global external @dora_global_ctx_is_static() {addr_space = 0 : i32} : !llvm.ptr
  func.func private @dora_fn_debug_print(i32)
  func.func private @dora_fn_write_result(!llvm.ptr, i64, i64, i64, i8) -> !llvm.ptr
  func.func private @dora_fn_keccak256_hasher(!llvm.ptr, i64, i64, !llvm.ptr) -> !llvm.ptr
  func.func private @dora_fn_calldata(!llvm.ptr) -> !llvm.ptr
  func.func private @dora_fn_calldata_size(!llvm.ptr) -> i64
  func.func private @dora_fn_chainid(!llvm.ptr) -> !llvm.ptr
  func.func private @dora_fn_callvalue(!llvm.ptr, !llvm.ptr) -> !llvm.ptr
  func.func private @dora_fn_caller(!llvm.ptr, !llvm.ptr) -> !llvm.ptr
  func.func private @dora_fn_store_in_gasprice_ptr(!llvm.ptr, !llvm.ptr) -> !llvm.ptr
  func.func private @dora_fn_store_in_selfbalance_ptr(!llvm.ptr, !llvm.ptr) -> !llvm.ptr
  func.func private @dora_fn_store_in_blobbasefee_ptr(!llvm.ptr, !llvm.ptr)
  func.func private @dora_fn_gaslimit(!llvm.ptr) -> !llvm.ptr
  func.func private @dora_fn_extend_memory(!llvm.ptr, i64) -> !llvm.ptr
  func.func private @dora_fn_code_copy(!llvm.ptr, i64, i64, i64) -> !llvm.ptr
  func.func private @dora_fn_sload(!llvm.ptr, !llvm.ptr, !llvm.ptr) -> !llvm.ptr
  func.func private @dora_fn_sstore(!llvm.ptr, !llvm.ptr, !llvm.ptr) -> !llvm.ptr
  func.func private @dora_fn_append_log(!llvm.ptr, i64, i64)
  func.func private @dora_fn_append_log_with_one_topic(!llvm.ptr, i64, i64, !llvm.ptr)
  func.func private @dora_fn_append_log_with_two_topics(!llvm.ptr, i64, i64, !llvm.ptr, !llvm.ptr)
  func.func private @dora_fn_append_log_with_three_topics(!llvm.ptr, i64, i64, !llvm.ptr, !llvm.ptr, !llvm.ptr)
  func.func private @dora_fn_append_log_with_four_topics(!llvm.ptr, i64, i64, !llvm.ptr, !llvm.ptr, !llvm.ptr, !llvm.ptr)
  func.func private @dora_fn_origin(!llvm.ptr, !llvm.ptr) -> !llvm.ptr
  func.func private @dora_fn_coinbase(!llvm.ptr) -> !llvm.ptr
  func.func private @dora_fn_block_number(!llvm.ptr, !llvm.ptr) -> !llvm.ptr
  func.func private @dora_fn_extcodesize(!llvm.ptr, !llvm.ptr) -> !llvm.ptr
  func.func private @dora_fn_address(!llvm.ptr) -> !llvm.ptr
  func.func private @dora_fn_prevrando(!llvm.ptr, !llvm.ptr) -> !llvm.ptr
  func.func private @dora_fn_store_in_timestamp_ptr(!llvm.ptr, !llvm.ptr) -> !llvm.ptr
  func.func private @dora_fn_store_in_basefee_ptr(!llvm.ptr, !llvm.ptr) -> !llvm.ptr
  func.func private @dora_fn_call(!llvm.ptr, i64, !llvm.ptr, !llvm.ptr, i64, i64, i64, i64, i64, !llvm.ptr, i8) -> !llvm.ptr
  func.func private @dora_fn_store_in_balance(!llvm.ptr, !llvm.ptr) -> !llvm.ptr
  func.func private @dora_fn_ext_code_copy(!llvm.ptr, !llvm.ptr, i64, i64, i64) -> !llvm.ptr
  func.func private @dora_fn_blob_hash(!llvm.ptr, !llvm.ptr)
  func.func private @dora_fn_block_hash(!llvm.ptr, !llvm.ptr) -> !llvm.ptr
  func.func private @dora_fn_ext_code_hash(!llvm.ptr, !llvm.ptr) -> !llvm.ptr
  func.func private @dora_fn_create(!llvm.ptr, i64, i64, !llvm.ptr, !llvm.ptr) -> !llvm.ptr
  func.func private @dora_fn_create2(!llvm.ptr, i64, i64, !llvm.ptr, !llvm.ptr, !llvm.ptr) -> !llvm.ptr
  func.func private @dora_fn_return_data_size(!llvm.ptr) -> !llvm.ptr
  func.func private @dora_fn_return_data_copy(!llvm.ptr, i64, i64, i64) -> !llvm.ptr
  func.func private @dora_fn_selfdestruct(!llvm.ptr, !llvm.ptr) -> !llvm.ptr
  func.func private @dora_fn_transient_storage_read(!llvm.ptr, !llvm.ptr, !llvm.ptr) -> !llvm.ptr
  func.func private @dora_fn_transient_storage_write(!llvm.ptr, !llvm.ptr, !llvm.ptr) -> !llvm.ptr
  func.func public @main(%arg0: !llvm.ptr, %arg1: i64) -> i8 attributes {llvm.emit_c_interface} {
    %c1024_i256 = arith.constant 1024 : i256
    %0 = llvm.alloca %c1024_i256 x i256 : (i256) -> !llvm.ptr
    %1 = llvm.mlir.addressof @dora_global_stack_ptr : !llvm.ptr
    llvm.store %0, %1 : !llvm.ptr, !llvm.ptr
    %c0_i64 = arith.constant 0 : i64
    %2 = llvm.mlir.addressof @dora_global_stack_length : !llvm.ptr
    llvm.store %c0_i64, %2 : i64, !llvm.ptr
    %c0_i64_0 = arith.constant 0 : i64
    %3 = llvm.mlir.addressof @dora_global_memory_size : !llvm.ptr
    llvm.store %c0_i64_0, %3 : i64, !llvm.ptr
    %4 = call @dora_fn_calldata(%arg0) : (!llvm.ptr) -> !llvm.ptr
    %5 = llvm.mlir.addressof @dora_global_calldata_ptr : !llvm.ptr
    llvm.store %4, %5 : !llvm.ptr, !llvm.ptr
    %6 = call @dora_fn_calldata_size(%arg0) : (!llvm.ptr) -> i64
    %7 = llvm.mlir.addressof @dora_global_calldata_size : !llvm.ptr
    llvm.store %6, %7 : i64, !llvm.ptr
    %8 = llvm.mlir.addressof @dora_global_gas_counter : !llvm.ptr
    llvm.store %arg1, %8 : i64, !llvm.ptr
    cf.br ^bb5
  ^bb1(%9: i8):  // 16 preds: ^bb2, ^bb4, ^bb5, ^bb8, ^bb9, ^bb12, ^bb13, ^bb15, ^bb16, ^bb17, ^bb18, ^bb19, ^bb20, ^bb21, ^bb23, ^bb24
    %c0_i64_1 = arith.constant 0 : i64
    %10 = llvm.mlir.addressof @dora_global_gas_counter : !llvm.ptr
    %11 = llvm.load %10 : !llvm.ptr -> i64
    %12 = call @dora_fn_write_result(%arg0, %c0_i64_1, %c0_i64_1, %11, %9) : (!llvm.ptr, i64, i64, i64, i8) -> !llvm.ptr
    return %9 : i8
  ^bb2(%13: i256):  // no predecessors
    %c89_i8 = arith.constant 89 : i8
    cf.switch %13 : i256, [
      default: ^bb1(%c89_i8 : i8)
    ]
  ^bb3:  // pred: ^bb4
    %c41_i256 = arith.constant 41 : i256
    %14 = llvm.mlir.addressof @dora_global_stack_ptr : !llvm.ptr
    %15 = llvm.load %14 : !llvm.ptr -> !llvm.ptr
    llvm.store %c41_i256, %15 : i256, !llvm.ptr
    %16 = llvm.getelementptr %15[1] : (!llvm.ptr) -> !llvm.ptr, i256
    llvm.store %16, %14 : !llvm.ptr, !llvm.ptr
    cf.br ^bb9
  ^bb4:  // pred: ^bb6
    %17 = llvm.mlir.addressof @dora_global_stack_length : !llvm.ptr
    %c1024_i64 = arith.constant 1024 : i64
    %18 = llvm.load %17 : !llvm.ptr -> i64
    %c1_i64 = arith.constant 1 : i64
    %19 = arith.addi %18, %c1_i64 : i64
    llvm.store %19, %17 : i64, !llvm.ptr
    %20 = arith.cmpi ult, %c1024_i64, %19 : i64
    %c92_i8 = arith.constant 92 : i8
    cf.cond_br %20, ^bb1(%c92_i8 : i8), ^bb3
  ^bb5:  // pred: ^bb0
    %21 = llvm.mlir.addressof @dora_global_gas_counter : !llvm.ptr
    %22 = llvm.load %21 : !llvm.ptr -> i64
    %c3_i64 = arith.constant 3 : i64
    %23 = arith.cmpi uge, %22, %c3_i64 : i64
    %c80_i8 = arith.constant 80 : i8
    cf.cond_br %23, ^bb6, ^bb1(%c80_i8 : i8)
  ^bb6:  // pred: ^bb5
    %24 = arith.subi %22, %c3_i64 : i64
    llvm.store %24, %21 : i64, !llvm.ptr
    cf.br ^bb4
  ^bb7:  // pred: ^bb8
    %c0_i256 = arith.constant 0 : i256
    %25 = llvm.mlir.addressof @dora_global_stack_ptr : !llvm.ptr
    %26 = llvm.load %25 : !llvm.ptr -> !llvm.ptr
    llvm.store %c0_i256, %26 : i256, !llvm.ptr
    %27 = llvm.getelementptr %26[1] : (!llvm.ptr) -> !llvm.ptr, i256
    llvm.store %27, %25 : !llvm.ptr, !llvm.ptr
    cf.br ^bb13
  ^bb8:  // pred: ^bb10
    %28 = llvm.mlir.addressof @dora_global_stack_length : !llvm.ptr
    %c1024_i64_2 = arith.constant 1024 : i64
    %29 = llvm.load %28 : !llvm.ptr -> i64
    %c1_i64_3 = arith.constant 1 : i64
    %30 = arith.addi %29, %c1_i64_3 : i64
    llvm.store %30, %28 : i64, !llvm.ptr
    %31 = arith.cmpi ult, %c1024_i64_2, %30 : i64
    %c92_i8_4 = arith.constant 92 : i8
    cf.cond_br %31, ^bb1(%c92_i8_4 : i8), ^bb7
  ^bb9:  // pred: ^bb3
    %32 = llvm.mlir.addressof @dora_global_gas_counter : !llvm.ptr
    %33 = llvm.load %32 : !llvm.ptr -> i64
    %c3_i64_5 = arith.constant 3 : i64
    %34 = arith.cmpi uge, %33, %c3_i64_5 : i64
    %c80_i8_6 = arith.constant 80 : i8
    cf.cond_br %34, ^bb10, ^bb1(%c80_i8_6 : i8)
  ^bb10:  // pred: ^bb9
    %35 = arith.subi %33, %c3_i64_5 : i64
    llvm.store %35, %32 : i64, !llvm.ptr
    cf.br ^bb8
  ^bb11:  // pred: ^bb12
    %c0_i256_7 = arith.constant 0 : i256
    %36 = llvm.mlir.addressof @dora_global_stack_ptr : !llvm.ptr
    %37 = llvm.load %36 : !llvm.ptr -> !llvm.ptr
    llvm.store %c0_i256_7, %37 : i256, !llvm.ptr
    %38 = llvm.getelementptr %37[1] : (!llvm.ptr) -> !llvm.ptr, i256
    llvm.store %38, %36 : !llvm.ptr, !llvm.ptr
    cf.br ^bb24
  ^bb12:  // pred: ^bb14
    %39 = llvm.mlir.addressof @dora_global_stack_length : !llvm.ptr
    %c1024_i64_8 = arith.constant 1024 : i64
    %40 = llvm.load %39 : !llvm.ptr -> i64
    %c1_i64_9 = arith.constant 1 : i64
    %41 = arith.addi %40, %c1_i64_9 : i64
    llvm.store %41, %39 : i64, !llvm.ptr
    %42 = arith.cmpi ult, %c1024_i64_8, %41 : i64
    %c92_i8_10 = arith.constant 92 : i8
    cf.cond_br %42, ^bb1(%c92_i8_10 : i8), ^bb11
  ^bb13:  // pred: ^bb7
    %43 = llvm.mlir.addressof @dora_global_gas_counter : !llvm.ptr
    %44 = llvm.load %43 : !llvm.ptr -> i64
    %c3_i64_11 = arith.constant 3 : i64
    %45 = arith.cmpi uge, %44, %c3_i64_11 : i64
    %c80_i8_12 = arith.constant 80 : i8
    cf.cond_br %45, ^bb14, ^bb1(%c80_i8_12 : i8)
  ^bb14:  // pred: ^bb13
    %46 = arith.subi %44, %c3_i64_11 : i64
    llvm.store %46, %43 : i64, !llvm.ptr
    cf.br ^bb12
  ^bb15:  // pred: ^bb23
    %47 = llvm.mlir.addressof @dora_global_stack_ptr : !llvm.ptr
    %48 = llvm.load %47 : !llvm.ptr -> !llvm.ptr
    %49 = llvm.getelementptr %48[-1] : (!llvm.ptr) -> !llvm.ptr, i256
    %50 = llvm.load %49 : !llvm.ptr -> i256
    llvm.store %49, %47 : !llvm.ptr, !llvm.ptr
    %51 = llvm.mlir.addressof @dora_global_stack_ptr : !llvm.ptr
    %52 = llvm.load %51 : !llvm.ptr -> !llvm.ptr
    %53 = llvm.getelementptr %52[-1] : (!llvm.ptr) -> !llvm.ptr, i256
    %54 = llvm.load %53 : !llvm.ptr -> i256
    llvm.store %53, %51 : !llvm.ptr, !llvm.ptr
    %55 = llvm.mlir.addressof @dora_global_stack_ptr : !llvm.ptr
    %56 = llvm.load %55 : !llvm.ptr -> !llvm.ptr
    %57 = llvm.getelementptr %56[-1] : (!llvm.ptr) -> !llvm.ptr, i256
    %58 = llvm.load %57 : !llvm.ptr -> i256
    llvm.store %57, %55 : !llvm.ptr, !llvm.ptr
    %59 = llvm.mlir.addressof @dora_global_ctx_is_static : !llvm.ptr
    %60 = llvm.load %59 : !llvm.ptr -> i1
    %c87_i8 = arith.constant 87 : i8
    cf.cond_br %60, ^bb1(%c87_i8 : i8), ^bb16
  ^bb16:  // pred: ^bb15
    %61 = arith.trunci %58 : i256 to i64
    %c0_i64_13 = arith.constant 0 : i64
    %62 = arith.cmpi slt, %61, %c0_i64_13 : i64
    %c84_i8 = arith.constant 84 : i8
    cf.cond_br %62, ^bb1(%c84_i8 : i8), ^bb17
  ^bb17:  // pred: ^bb16
    %c49152_i64 = arith.constant 49152 : i64
    %63 = arith.cmpi ugt, %61, %c49152_i64 : i64
    %c100_i8 = arith.constant 100 : i8
    cf.cond_br %63, ^bb1(%c100_i8 : i8), ^bb18
  ^bb18:  // pred: ^bb17
    %64 = llvm.mlir.addressof @dora_global_gas_counter : !llvm.ptr
    %65 = llvm.load %64 : !llvm.ptr -> i64
    %66 = arith.trunci %58 : i256 to i64
    %c0_i64_14 = arith.constant 0 : i64
    %67 = arith.cmpi eq, %66, %c0_i64_14 : i64
    %68 = scf.if %67 -> (i64) {
      %c0_i64_25 = arith.constant 0 : i64
      scf.yield %c0_i64_25 : i64
    } else {
      %c2_i64 = arith.constant 2 : i64
      %c31_i64_25 = arith.constant 31 : i64
      %c32_i64_26 = arith.constant 32 : i64
      %109 = arith.addi %66, %c31_i64_25 : i64
      %110 = arith.divui %109, %c32_i64_26 : i64
      %111 = arith.muli %110, %c2_i64 : i64
      scf.yield %111 : i64
    }
    %69 = arith.cmpi uge, %65, %68 : i64
    %c80_i8_15 = arith.constant 80 : i8
    cf.cond_br %69, ^bb27, ^bb1(%c80_i8_15 : i8)
  ^bb19:  // pred: ^bb27
    %70 = llvm.mlir.addressof @dora_global_gas_counter : !llvm.ptr
    %71 = llvm.load %70 : !llvm.ptr -> i64
    %72 = arith.trunci %58 : i256 to i64
    %c0_i64_16 = arith.constant 0 : i64
    %73 = arith.cmpi eq, %72, %c0_i64_16 : i64
    %74 = scf.if %73 -> (i64) {
      %c0_i64_25 = arith.constant 0 : i64
      scf.yield %c0_i64_25 : i64
    } else {
      %109 = arith.trunci %54 : i256 to i64
      %110 = arith.addi %109, %72 : i64
      %111 = llvm.mlir.addressof @dora_global_memory_size : !llvm.ptr
      %112 = llvm.load %111 : !llvm.ptr -> i64
      %c31_i64_25 = arith.constant 31 : i64
      %c32_i64_26 = arith.constant 32 : i64
      %113 = arith.addi %110, %c31_i64_25 : i64
      %114 = arith.divui %113, %c32_i64_26 : i64
      %115 = arith.muli %114, %c32_i64_26 : i64
      %116 = arith.cmpi ult, %112, %115 : i64
      %117 = scf.if %116 -> (i64) {
        %c3_i64_27 = arith.constant 3 : i64
        %c31_i64_28 = arith.constant 31 : i64
        %c32_i64_29 = arith.constant 32 : i64
        %c512_i64 = arith.constant 512 : i64
        %118 = arith.addi %112, %c31_i64_28 : i64
        %119 = arith.divui %118, %c32_i64_29 : i64
        %120 = arith.muli %119, %119 : i64
        %121 = arith.divui %120, %c512_i64 : i64
        %122 = arith.muli %119, %c3_i64_27 : i64
        %123 = arith.addi %121, %122 : i64
        %c3_i64_30 = arith.constant 3 : i64
        %c31_i64_31 = arith.constant 31 : i64
        %c32_i64_32 = arith.constant 32 : i64
        %c512_i64_33 = arith.constant 512 : i64
        %124 = arith.addi %115, %c31_i64_31 : i64
        %125 = arith.divui %124, %c32_i64_32 : i64
        %126 = arith.muli %125, %125 : i64
        %127 = arith.divui %126, %c512_i64_33 : i64
        %128 = arith.muli %125, %c3_i64_30 : i64
        %129 = arith.addi %127, %128 : i64
        %130 = arith.subi %129, %123 : i64
        scf.yield %130 : i64
      } else {
        %c0_i64_27 = arith.constant 0 : i64
        scf.yield %c0_i64_27 : i64
      }
      scf.yield %117 : i64
    }
    %75 = arith.cmpi uge, %71, %74 : i64
    %c80_i8_17 = arith.constant 80 : i8
    cf.cond_br %75, ^bb28, ^bb1(%c80_i8_17 : i8)
  ^bb20:  // pred: ^bb28
    %76 = llvm.mlir.addressof @dora_global_gas_counter : !llvm.ptr
    %77 = llvm.load %76 : !llvm.ptr -> i64
    %78 = arith.trunci %58 : i256 to i64
    %c32000_i64 = arith.constant 32000 : i64
    %c6_i64 = arith.constant 6 : i64
    %c31_i64 = arith.constant 31 : i64
    %c32_i64 = arith.constant 32 : i64
    %79 = arith.addi %78, %c31_i64 : i64
    %80 = arith.divui %79, %c32_i64 : i64
    %81 = arith.muli %80, %c6_i64 : i64
    %82 = arith.addi %c32000_i64, %81 : i64
    %83 = arith.cmpi uge, %77, %82 : i64
    %c80_i8_18 = arith.constant 80 : i8
    cf.cond_br %83, ^bb29, ^bb1(%c80_i8_18 : i8)
  ^bb21:  // pred: ^bb29
    %84 = llvm.mlir.addressof @dora_global_gas_counter : !llvm.ptr
    %85 = llvm.load %84 : !llvm.ptr -> i64
    %86 = llvm.mlir.addressof @dora_global_gas_counter : !llvm.ptr
    %87 = llvm.load %86 : !llvm.ptr -> i64
    %c64_i64 = arith.constant 64 : i64
    %88 = arith.divui %87, %c64_i64 : i64
    %89 = arith.cmpi uge, %85, %88 : i64
    %c80_i8_19 = arith.constant 80 : i8
    cf.cond_br %89, ^bb30, ^bb1(%c80_i8_19 : i8)
  ^bb22:  // pred: ^bb30
    %90 = "dora.create"(%50, %54, %58) : (i256, i256, i256) -> i256
    %91 = llvm.mlir.addressof @dora_global_stack_ptr : !llvm.ptr
    %92 = llvm.load %91 : !llvm.ptr -> !llvm.ptr
    llvm.store %90, %92 : i256, !llvm.ptr
    %93 = llvm.getelementptr %92[1] : (!llvm.ptr) -> !llvm.ptr, i256
    llvm.store %93, %91 : !llvm.ptr, !llvm.ptr
    cf.br ^bb26
  ^bb23:  // pred: ^bb25
    %94 = llvm.mlir.addressof @dora_global_stack_length : !llvm.ptr
    %c1024_i64_20 = arith.constant 1024 : i64
    %95 = llvm.load %94 : !llvm.ptr -> i64
    %c-2_i64 = arith.constant -2 : i64
    %96 = arith.addi %95, %c-2_i64 : i64
    llvm.store %96, %94 : i64, !llvm.ptr
    %c3_i64_21 = arith.constant 3 : i64
    %97 = arith.cmpi ult, %95, %c3_i64_21 : i64
    %c91_i8 = arith.constant 91 : i8
    cf.cond_br %97, ^bb1(%c91_i8 : i8), ^bb15
  ^bb24:  // pred: ^bb11
    %98 = llvm.mlir.addressof @dora_global_gas_counter : !llvm.ptr
    %99 = llvm.load %98 : !llvm.ptr -> i64
    %c0_i64_22 = arith.constant 0 : i64
    %100 = arith.cmpi uge, %99, %c0_i64_22 : i64
    %c80_i8_23 = arith.constant 80 : i8
    cf.cond_br %100, ^bb25, ^bb1(%c80_i8_23 : i8)
  ^bb25:  // pred: ^bb24
    %101 = arith.subi %99, %c0_i64_22 : i64
    llvm.store %101, %98 : i64, !llvm.ptr
    cf.br ^bb23
  ^bb26:  // pred: ^bb22
    %c0_i64_24 = arith.constant 0 : i64
    %c1_i8 = arith.constant 1 : i8
    %102 = llvm.mlir.addressof @dora_global_gas_counter : !llvm.ptr
    %103 = llvm.load %102 : !llvm.ptr -> i64
    %104 = call @dora_fn_write_result(%arg0, %c0_i64_24, %c0_i64_24, %103, %c1_i8) : (!llvm.ptr, i64, i64, i64, i8) -> !llvm.ptr
    return %c1_i8 : i8
  ^bb27:  // pred: ^bb18
    %105 = arith.subi %65, %68 : i64
    llvm.store %105, %64 : i64, !llvm.ptr
    cf.br ^bb19
  ^bb28:  // pred: ^bb19
    %106 = arith.subi %71, %74 : i64
    llvm.store %106, %70 : i64, !llvm.ptr
    cf.br ^bb20
  ^bb29:  // pred: ^bb20
    %107 = arith.subi %77, %82 : i64
    llvm.store %107, %76 : i64, !llvm.ptr
    cf.br ^bb21
  ^bb30:  // pred: ^bb21
    %108 = arith.subi %85, %88 : i64
    llvm.store %108, %84 : i64, !llvm.ptr
    cf.br ^bb22
  }
}
